Migrate task listener

smbea opened the pull request:
Closes #106

barmac commented on src/provider/camunda-platform/properties/ListenerProps.js:
```diff
@@ -366,6 +478,14 @@ function AddListener(props) {
   );
 }
 
+function AddTaskListener(props) {
+  return AddListener({ ...props, listenerGroup:'camunda:TaskListener' });
+}
+
+function AddExecutionListener(props) {
+  return AddListener({ ...props, listenerGroup:'camunda:ExecutionListener' });
+}
+
```
I like it.

smbea commented on src/provider/camunda-platform/properties/ListenerProps.js:
```diff
@@ -150,6 +157,68 @@ function ExecutionListener(props) {
+
+  return {
+    items: listeners.map((listener, index) => {
+      const id = `${element.id}-listener-${index}`;
```
Should we do `${element.id}-executionlistener-${index}` or `${element.id}-tasklistener-${index}` then?

barmac replied:
It could be ...-executionListener-... and ...-taskListener-..., but the most important is that the ids are different.

smbea replied:
I updated the ids

barmac commented on src/provider/camunda-platform/properties/ListenerProps.js:
```diff
@@ -195,6 +264,18 @@ function EventType({ id, element, listener }) {
 
   function getOptions() {
+
+    if (listener.$type == 'camunda:TaskListener') {
```
```suggestion
    if (is(listener, 'camunda:TaskListener')) {
```

We don't want to depend on internals.

barmac commented on src/provider/camunda-platform/utils/ImplementationTypeUtils.js:
```diff
@@ -142,7 +142,7 @@ export function getImplementationType(element) {
 
 function getListenerBusinessObject(businessObject) {
-  if (is(businessObject, 'camunda:ExecutionListener')) {
+  if (is(businessObject, 'camunda:ExecutionListener') || is(businessObject,'camunda:TaskListener')) {
```
There will be probably no more types of listeners, but generally we could use `isAny` for multiple types check.

barmac commented:
Just one more comment to address and we are ready to merge this.

andreasgeier reviewed (changes requested):
The UI looks as expected.

But functional-wise, I miss the timer-related properties (type + definition) when creating a task listener of type timeout.

smbea commented:
@barmac I made the changes to use `is` and `isAny`
@andreasgeier I added the timer related properties:

<img width="234" alt="Screenshot 2021-10-05 at 10 20 09" src="https://user-images.githubusercontent.com/25825387/135986717-ba1ec0c5-1047-403a-ae7d-8f2ce7c48908.png">

This resulted in some changes since the last review, specifically in the TimerProps and ListenerProps

andreasgeier reviewed (approved):
✅ &nbsp; LGTM

barmac commented on src/provider/camunda-platform/properties/ListenerProps.js:
```diff
@@ -179,22 +251,37 @@ function RemoveListenerContainer(props) {
+
+    commandStack.execute('properties-panel.update-businessobject', {
+      element: element,
+      businessObject: getBusinessObject(listener),
```
Listener is always a business object, so there's no need to call the helper here.

Generally speaking, we have two types of complex objects in bpmn-js:
* diagram-js objects: shapes, connections -> which have a property called `businessObject` linking to the second type
* bpmn-js business objects: bpmn:Event-s, bpmn:Task-s and so on -> these are objects which are serialized; they don't have `businessObject` ever

barmac commented on src/provider/camunda-platform/properties/ListenerProps.js:
```diff
@@ -410,6 +554,15 @@ function getDefaultImplementationProperties(type, bpmnFactory) {
+  case 'timeout':
+    return { eventDefinitions: [bpmnFactory.create('bpmn:TimerEventDefinition')], event:type };
+  default:
+    return { event: type };
```
Note that we do not clean up the event definitions when we change from `timeout` event to something else. Please add a test case that verifies this. It could be a set of actions: create -> timeout -> assignment.

smbea replied:
I cleared the definitions when the type is changed from 'timeout' to something else and added a test case to verify

barmac commented on src/provider/camunda-platform/properties/ListenerProps.js:
```diff
@@ -214,6 +301,36 @@ function EventType({ id, element, listener }) {
+  const translate = useService('translate');
+  const debounce = useService('debounceInput');
+  const commandStack = useService('commandStack');
+  const businessObject = getBusinessObject(listener);
```
Again, listener is always a business object.

barmac reviewed (changes requested):
Good job overall. I left a couple of comments which need to be addressed before we merge this.

barmac commented:
It works fine. The time out problem should disappear when you rebase on `main`.
